### ISSUE TITLE
Key pressed not working for alphanumeric keys on Microsoft Edge

### ISSUE THREAD

**Issue by jorenl** [LLK/scratch-flash#1334]
Hey everyone,

I ran into this issue with a project I'm building, and then found a forum thread on the issue:
https://scratch.mit.edu/discuss/topic/251930

It seems very serious. The <when key [... v] pressed?> blocks returns false, always, for any letter or number keys. Space bar and arrow keys seem to work fine. Key pressed event hats also don't trigger.

I shared this project to test this:
https://scratch.mit.edu/projects/170587322/
I can press all keys fine in Chrome, but only the arrows and the space bar in Edge.

Entering text in the editor, like for changing text values, works just fine. The issue only seems to affect projects.

Running Microsoft Edge 40.15063.0.0
Windows Version 10.0.15063 Build 15063

**Comment by thisandagain** [comment]
Thanks for reporting @jorenl. Agreed that this seems serious. We'll investigate.

**Comment by thisandagain** [comment]
@jwzimmer Would you mind attempting to repro?

**Comment by jwzimmer** [comment]
### Repro

Edge 40.15063.0.0 (Same as reporting user)
Windows 10 Build 15063 (MSEdge - Win10_preview VM on qa-machine)

### Repro attempts in various browsers/ OS's
- **Repro'd in Edge on Win 10, not in IE 11 on Win 10 or Chrome on Mac OS**
- Edge (on Windows VM): Using project 170587322
  - The arrow & space costumes do grow in size in response to being pressed, as expected
  - The number & letter keys do not seem to cause a reaction
  - No errors in the console
- Internet Explorer 11 (on Windows VM): Using project 170587322
  - The arrow & space costumes do grow in size in response to being pressed, as expected
  - The number & letter keys do grow in size in response to being pressed, as expected
- Chrome 60.0.3112.78 on Mac OS El Capitan: Using project 170587322
  - The arrow & space costumes do grow in size in response to being pressed, as expected
  - The number & letter keys do grow in size in response to being pressed, as expected

### Steps to repro using project 170587322
- **Shortest steps to repro**
- Open Edge on the qa-machine Win 10 preview VM
- Go to https://scratch.mit.edu/projects/170587322/
- Press the green flag
- Press the arrow or space keys, note the change in size of the respective costumes on the stage
- Press any letter or number keys, not the lack of change in size of the respective costumes on the stage

### Steps to repro from a new/ blank project
- **Hat block repro**
- Open Edge on the qa-machine Win 10 preview VM
- Go to scratch homepage
- Click Create
- Drag out the hat block "When space key pressed"
- Connect to it the block "change color effect by 25"
- Click the stack to trigger it, note the Scratch cat changes color
- Press space to trigger the stack, note the Scratch cat changes color
- Change "space" to be "a" in the hat block
- Click the stack to trigger it, note the Scratch cat changes color
- Press the "a" key to trigger the stack, note the Scratch cat does not change color
  - There are no errors in the console
- **Sensing block repro**
- Drag out the "if" control loop
- Put the "change color effect by 25" block inside the if-loop
- Insert the "key space pressed?" sensing block in the if-statement
- Hold down the space key
- Click the "if" stack, note the Scratch cat changes color
- Change the "space" to be "a"
- Hold down the "a" key
- Click the "if" stack, note the Scratch cat does not change color

**Comment by cwillisf** [comment]
The problem is that the [`KeyboardEvent` in Flash](http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/events/KeyboardEvent.html) is reporting a value of 0 for the `charCode`. This appears to be a genuine bug in Edge's bundled implementation of Flash.

I've also put together a workaround based on reading the `keyCode` property instead of `charCode`. It needs a little more testing but so far it seems reliable in all browsers.

**Comment by jwzimmer** [comment]
### Testing

1) This stack seemed to crash Flash Player in Chrome, Windows 10, @cwillisf's build, such that I couldn't stop the script by clicking the red flag (I tried several times), but this only happened once. 
<img width="679" alt="screen shot 2017-08-15 at 4 55 14 pm" src="https://user-images.githubusercontent.com/1320456/29336083-c09ad034-81da-11e7-903d-2493055cca98.png">

2) I still notice that if I repeatedly press the space bar, the cat's color changing doesn't keep up as well with the key presses in Edge as in Chrome (@cwillisf's build or Staging). The stack I'm using is:
- when space key pressed
- forever
- if key space pressed? then
- change color effect by 25

I think something bad is still happening that causes a backlog of key presses to build up - I think 1 & 2 are related. I just had a delay in the colors changing, then none of my clicks to the red stop sign seemed to register, then I saw the red stop sign change state several times (as if registering the clicks after a delay), then the program stopped.

**Comment by cwillisf** [comment]
@jwzimmer I was unable to reproduce any of the issues from your [report](https://github.com/LLK/scratch-flash/issues/1334#issuecomment-322588054)... however I do have some thoughts.

1. I've noticed Chrome pausing for a few seconds at a time on my machine, starting yesterday. It doesn't seem to be related to Scratch or even Flash: in fact, it seemed to happen most often when I was editing tips files, which just involved reloading plain HTML files a bunch. I'm trying to narrow down what's causing it, but it's intermittent so it could take a while. It's also possible it'll turn out to just be a weird thing my computer is doing, unrelated to what you saw.

2. For me it seems like Chrome is very slightly slower than Edge if I hold the space key down, but they both respond equally if I tap the space key. I wonder if it's a matter of the two applications processing the key repeat rate differently...

3. When I was trying to repro this, I failed at first because my "when space key pressed" hat was still glowing from when it had the "forever" loop attached. Most hat blocks don't "retrigger" in this type of situation, so I had to toggle it off or press the stop button to get it to work again. Any chance this is what happened to you in Edge?
